File: tools/regiondb-build/src/regiondb_build/blob.py
"""The compiled geometry encoding.

Polygons are stored as a project-defined blob rather than WKB or SpatiaLite
geometry, for two reasons: the runtime must not need a SQLite extension on
iOS, and coordinates quantized to a fixed integer grid make boundary tests
exact instead of epsilon-dependent. One blob holds one connected polygon
component — an exterior ring plus its holes — so a country's remote islands
each get their own bounding box and the R-tree stays useful.

Coordinates are 1e-6 degrees (about 11 cm at the equator), delta-encoded from
the previous vertex and zigzag-varint packed. Ring closure is implicit; the
repeated final vertex is never stored.
"""

from __future__ import annotations

from dataclasses import dataclass

GEOMETRY_FORMAT_VERSION = 1

RING_EXTERIOR = 0
RING_HOLE = 1

COORD_SCALE = 1_000_000


class GeometryBlobError(ValueError):
    """A blob is malformed or uses an unsupported version."""


@dataclass(frozen=True)
class Ring:
    """One closed ring as integer 1e-6-degree coordinates, without closure."""

    role: int
    points: tuple[tuple[int, int], ...]


def to_e6(degrees: float) -> int:
    """Quantize a degree value onto the storage grid.

    Round-half-away-from-zero rather than Python's banker's rounding, so the
    Rust and JavaScript readers can reproduce the grid with plain arithmetic.
    """
    scaled = degrees * COORD_SCALE
    return int(scaled + 0.5) if scaled >= 0 else -int(-scaled + 0.5)


def from_e6(value: int) -> float:
    return value / COORD_SCALE


def _write_varint(out: bytearray, value: int) -> None:
    if value < 0:
        raise GeometryBlobError("varint values are unsigned")
    while True:
        byte = value & 0x7F
        value >>= 7
        if value:
            out.append(byte | 0x80)
        else:
            out.append(byte)
            return


def _write_zigzag(out: bytearray, value: int) -> None:
    _write_varint(out, ((-value) << 1) - 1 if value < 0 else value << 1)


def _read_varint(data: bytes, offset: int) -> tuple[int, int]:
    result = 0
    shift = 0
    while True:
        if offset >= len(data):
            raise GeometryBlobError("truncated varint")
        byte = data[offset]
        offset += 1
        result |= (byte & 0x7F) << shift
        if not byte & 0x80:
            return result, offset
        shift += 7
        if shift > 63:
            raise GeometryBlobError("varint too long")


def _read_zigzag(data: bytes, offset: int) -> tuple[int, int]:
    raw, offset = _read_varint(data, offset)
    return (raw >> 1) ^ -(raw & 1), offset


def encode(rings: list[Ring]) -> bytes:
    """Encode one polygon component."""
    if not rings:
        raise GeometryBlobError("a geometry part needs at least one ring")
    if rings[0].role != RING_EXTERIOR:
        raise GeometryBlobError("the first ring of a part must be its exterior")

    out = bytearray()
    out.append(GEOMETRY_FORMAT_VERSION)
    _write_varint(out, len(rings))
    for ring in rings:
        if len(ring.points) < 3:
            raise GeometryBlobError("a ring needs at least three distinct vertices")
        out.append(ring.role)
        _write_varint(out, len(ring.points))
        previous_lon = 0
        previous_lat = 0
        for lon, lat in ring.points:
            _write_zigzag(out, lon - previous_lon)
            _write_zigzag(out, lat - previous_lat)
            previous_lon = lon
            previous_lat = lat
    return bytes(out)


def decode(blob: bytes) -> list[Ring]:
    """Decode one polygon component."""
    if not blob:
        raise GeometryBlobError("empty geometry blob")
    version = blob[0]
    if version != GEOMETRY_FORMAT_VERSION:
        raise GeometryBlobError(f"unsupported geometry format version {version}")

    offset = 1
    ring_count, offset = _read_varint(blob, offset)
    rings: list[Ring] = []
    for _ in range(ring_count):
        if offset >= len(blob):
            raise GeometryBlobError("truncated ring header")
        role = blob[offset]
        offset += 1
        point_count, offset = _read_varint(blob, offset)
        points: list[tuple[int, int]] = []
        lon = 0
        lat = 0
        for _ in range(point_count):
            delta_lon, offset = _read_zigzag(blob, offset)
            delta_lat, offset = _read_zigzag(blob, offset)
            lon += delta_lon
            lat += delta_lat
            points.append((lon, lat))
        rings.append(Ring(role=role, points=tuple(points)))
    if offset != len(blob):
        raise GeometryBlobError("trailing bytes after geometry blob")
    return rings


def point_in_rings(rings: list[Ring], lon_e6: int, lat_e6: int) -> bool:
    """Exact integer point-in-polygon, boundary inclusive.

    A point exactly on a boundary counts as inside — everywhere, in every
    implementation. Two abutting regions therefore both claim their shared
    edge, which is the harmless failure; a gap between them would leave a
    position with no region at all.
    """
    exterior = [ring for ring in rings if ring.role == RING_EXTERIOR]
    holes = [ring for ring in rings if ring.role == RING_HOLE]
    if not exterior:
        return False

    for ring in rings:
        if _on_ring(ring, lon_e6, lat_e6):
            return True
    if not _strictly_inside(exterior[0], lon_e6, lat_e6):
        return False
    return not any(_strictly_inside(hole, lon_e6, lat_e6) for hole in holes)


def _on_ring(ring: Ring, x: int, y: int) -> bool:
    points = ring.points
    for index in range(len(points)):
        x1, y1 = points[index]
        x2, y2 = points[(index + 1) % len(points)]
        if (x2 - x1) * (y - y1) - (y2 - y1) * (x - x1) != 0:
            continue
        if min(x1, x2) <= x <= max(x1, x2) and min(y1, y2) <= y <= max(y1, y2):
            return True
    return False


def _strictly_inside(ring: Ring, x: int, y: int) -> bool:
    """Crossing-number test with a half-open rule on the vertical span.

    Treating each edge as containing its lower endpoint and not its upper one
    makes a ray through a vertex count once rather than twice or zero times,
    with no tolerance anywhere.
    """
    points = ring.points
    inside = False
    for index in range(len(points)):
        x1, y1 = points[index]
        x2, y2 = points[(index + 1) % len(points)]
        if (y1 > y) != (y2 > y):
            # Sign-safe form of `x < x1 + (y - y1) * (x2 - x1) / (y2 - y1)`.
            side = (x2 - x1) * (y - y1) - (y2 - y1) * (x - x1)
            if side != 0 and (side > 0) == (y2 > y1):
                inside = not inside
    return inside

File: crates/umsh-regiondb/src/blob.rs
//! The compiled geometry encoding.
//!
//! One blob holds one connected polygon component: an exterior ring and its
//! holes, as 1e-6-degree integer coordinates, delta-encoded and zigzag-varint
//! packed, with ring closure implicit.
//!
//! Working in fixed-point integers is what makes the boundary rule exact. A
//! point on a boundary counts as inside, everywhere, and deciding that with
//! floating-point coordinates would mean choosing an epsilon and hoping three
//! implementations chose the same one.

/// Version of the blob layout this reader understands.
pub const GEOMETRY_FORMAT_VERSION: u8 = 1;

/// Coordinate units per degree.
pub const COORD_SCALE: f64 = 1_000_000.0;

/// The role a ring plays in its polygon.
pub const RING_EXTERIOR: u8 = 0;
/// A hole cut out of the exterior.
pub const RING_HOLE: u8 = 1;

/// A malformed or unsupported geometry blob.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum GeometryError {
    /// The blob was empty.
    Empty,
    /// The blob declares a version this reader does not implement.
    UnsupportedVersion(u8),
    /// The blob ended in the middle of a value.
    Truncated,
    /// A varint ran past 64 bits.
    VarintTooLong,
    /// Bytes remained after the last ring.
    TrailingBytes,
    /// The first ring was not an exterior.
    MissingExterior,
}

impl core::fmt::Display for GeometryError {
    fn fmt(&self, formatter: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::Empty => write!(formatter, "empty geometry blob"),
            Self::UnsupportedVersion(version) => {
                write!(formatter, "unsupported geometry format version {version}")
            }
            Self::Truncated => write!(formatter, "truncated geometry blob"),
            Self::VarintTooLong => write!(formatter, "varint too long"),
            Self::TrailingBytes => write!(formatter, "trailing bytes after geometry blob"),
            Self::MissingExterior => write!(formatter, "geometry part has no exterior ring"),
        }
    }
}

impl std::error::Error for GeometryError {}

/// One closed ring, without its repeated closing vertex.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Ring {
    /// [`RING_EXTERIOR`] or [`RING_HOLE`].
    pub role: u8,
    /// Vertices as `(longitude, latitude)` in 1e-6 degrees.
    pub points: Vec<(i32, i32)>,
}

/// Quantize a degree value onto the storage grid, rounding half away from zero.
pub fn to_e6(degrees: f64) -> i32 {
    let scaled = degrees * COORD_SCALE;
    let rounded = if scaled >= 0.0 {
        (scaled + 0.5).floor()
    } else {
        -((-scaled + 0.5).floor())
    };
    rounded.clamp(i32::MIN as f64, i32::MAX as f64) as i32
}

/// Convert a stored coordinate back to degrees.
pub fn from_e6(value: i32) -> f64 {
    f64::from(value) / COORD_SCALE
}

struct Reader<'a> {
    data: &'a [u8],
    offset: usize,
}

impl<'a> Reader<'a> {
    fn byte(&mut self) -> Result<u8, GeometryError> {
        let byte = *self.data.get(self.offset).ok_or(GeometryError::Truncated)?;
        self.offset += 1;
        Ok(byte)
    }

    fn varint(&mut self) -> Result<u64, GeometryError> {
        let mut result: u64 = 0;
        let mut shift = 0u32;
        loop {
            let byte = self.byte()?;
            result |= u64::from(byte & 0x7F) << shift;
            if byte & 0x80 == 0 {
                return Ok(result);
            }
            shift += 7;
            if shift > 63 {
                return Err(GeometryError::VarintTooLong);
            }
        }
    }

    fn zigzag(&mut self) -> Result<i64, GeometryError> {
        let raw = self.varint()?;
        Ok(((raw >> 1) as i64) ^ -((raw & 1) as i64))
    }
}

/// Decode one polygon component.
pub fn decode(data: &[u8]) -> Result<Vec<Ring>, GeometryError> {
    if data.is_empty() {
        return Err(GeometryError::Empty);
    }
    let version = data[0];
    if version != GEOMETRY_FORMAT_VERSION {
        return Err(GeometryError::UnsupportedVersion(version));
    }

    let mut reader = Reader { data, offset: 1 };
    let ring_count = reader.varint()?;
    let mut rings = Vec::with_capacity(ring_count as usize);
    for index in 0..ring_count {
        let role = reader.byte()?;
        if index == 0 && role != RING_EXTERIOR {
            return Err(GeometryError::MissingExterior);
        }
        let point_count = reader.varint()?;
        let mut points = Vec::with_capacity(point_count as usize);
        let mut longitude: i64 = 0;
        let mut latitude: i64 = 0;
        for _ in 0..point_count {
            longitude += reader.zigzag()?;
            latitude += reader.zigzag()?;
            points.push((longitude as i32, latitude as i32));
        }
        rings.push(Ring { role, points });
    }
    if reader.offset != data.len() {
        return Err(GeometryError::TrailingBytes);
    }
    Ok(rings)
}

/// Exact integer point-in-polygon, boundary inclusive.
///
/// A point exactly on a boundary is inside. Two abutting regions therefore both
/// claim their shared edge, which costs an operator one extra region in a list
/// they are reviewing anyway; a gap between them would leave a position with no
/// region at all.
pub fn point_in_rings(rings: &[Ring], longitude_e6: i32, latitude_e6: i32) -> bool {
    let mut exterior = None;
    for ring in rings {
        if on_ring(ring, longitude_e6, latitude_e6) {
            return true;
        }
        if ring.role == RING_EXTERIOR && exterior.is_none() {
            exterior = Some(ring);
        }
    }

    let Some(exterior) = exterior else {
        return false;
    };
    if !strictly_inside(exterior, longitude_e6, latitude_e6) {
        return false;
    }
    !rings
        .iter()
        .filter(|ring| ring.role == RING_HOLE)
        .any(|hole| strictly_inside(hole, longitude_e6, latitude_e6))
}

fn on_ring(ring: &Ring, x: i32, y: i32) -> bool {
    let points = &ring.points;
    let x = i64::from(x);
    let y = i64::from(y);
    for index in 0..points.len() {
        let (x1, y1) = points[index];
        let (x2, y2) = points[(index + 1) % points.len()];
        let (x1, y1, x2, y2) = (i64::from(x1), i64::from(y1), i64::from(x2), i64::from(y2));
        if (x2 - x1) * (y - y1) - (y2 - y1) * (x - x1) != 0 {
            continue;
        }
        if x1.min(x2) <= x && x <= x1.max(x2) && y1.min(y2) <= y && y <= y1.max(y2) {
            return true;
        }
    }
    false
}

/// Crossing-number test with a half-open rule on each edge's vertical span, so
/// that a ray through a vertex counts once rather than twice or not at all.
fn strictly_inside(ring: &Ring, x: i32, y: i32) -> bool {
    let points = &ring.points;
    let x = i64::from(x);
    let y = i64::from(y);
    let mut inside = false;
    for index in 0..points.len() {
        let (x1, y1) = points[index];
        let (x2, y2) = points[(index + 1) % points.len()];
        let (x1, y1, x2, y2) = (i64::from(x1), i64::from(y1), i64::from(x2), i64::from(y2));
        if (y1 > y) != (y2 > y) {
            let side = (x2 - x1) * (y - y1) - (y2 - y1) * (x - x1);
            if side != 0 && (side > 0) == (y2 > y1) {
                inside = !inside;
            }
        }
    }
    inside
}
